### PR TITLE
Add a rnn example on mnist dataset using tf library

### PR DESCRIPTION
**Pull request by imsheridan**
A Recurrent Neural Network (LSTM) implementation example using TensorFlow library.
This example is using the MNIST database of handwritten digits (http://yann.lecun.com/exdb/mnist/)
Links:
    [Long Short Term Memory](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf)
    [MNIST Dataset](http://yann.lecun.com/exdb/mnist/).

Training and evaluation log:
Extracting /tmp/mnist_data/train-images-idx3-ubyte.gz
Extracting /tmp/mnist_data/train-labels-idx1-ubyte.gz
Extracting /tmp/mnist_data/t10k-images-idx3-ubyte.gz
Extracting /tmp/mnist_data/t10k-labels-idx1-ubyte.gz
Step 1, Minibatch Loss= 2.5586, Training Accuracy= 0.258
Step 200, Minibatch Loss= 0.2419, Training Accuracy= 0.930
Step 400, Minibatch Loss= 0.1863, Training Accuracy= 0.938
Step 600, Minibatch Loss= 0.1000, Training Accuracy= 0.969
Step 800, Minibatch Loss= 0.0935, Training Accuracy= 0.977
Step 1000, Minibatch Loss= 0.0773, Training Accuracy= 0.969
Step 1200, Minibatch Loss= 0.0500, Training Accuracy= 0.984
Step 1400, Minibatch Loss= 0.0550, Training Accuracy= 0.977
Step 1600, Minibatch Loss= 0.0615, Training Accuracy= 0.984
Step 1800, Minibatch Loss= 0.0635, Training Accuracy= 0.977
Step 2000, Minibatch Loss= 0.0550, Training Accuracy= 0.992
Training Finished!
Testing Accuracy: 0.992188

### CODE REVIEW

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+""" Recurrent Neural Network.
+A Recurrent Neural Network (LSTM) implementation example using TensorFlow library.
+This example is using the MNIST database of handwritten digits (http://yann.lecun.com/exdb/mnist/)
+Links:
+    [Long Short Term Memory](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf)
+    [MNIST Dataset](http://yann.lecun.com/exdb/mnist/).
+"""
+
+from __future__ import print_function
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Add the other futures (`division`, `absolute_import`).

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+""" Recurrent Neural Network.
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: I think you need the copyright office.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    [Long Short Term Memory](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf)
+    [MNIST Dataset](http://yann.lecun.com/exdb/mnist/).
+"""
+
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.contrib import rnn
+
+# Import MNIST data
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Punctuation for comments: here and below.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+from tensorflow.examples.tutorials.mnist import input_data
+mnist = input_data.read_data_sets("/tmp/mnist_data/", one_hot=True)
+
+'''
+To classify images using a recurrent neural network, we consider every image
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: maybe put that as a comment

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+display_step = 200
+
+# Network Parameters
+num_input = 28 # MNIST data input (img shape: 28*28)
+num_step = 28 # step number
+num_hidden_units = 128 # hidden layer num of features
+num_classes = 10 # MNIST total classes (0-9 digits)
+
+# tf Graph input
+X = tf.placeholder("float", [None, num_step, num_input])
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Please use consistent quote, it sounds like you use single quote mostly (`'`).

Comment (drpngx) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Also, maybe use `shape=`. Also use something more informative, like `INPUT_IMAGES` and `OUTPUT_CLASSES`.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+biases = {
+    # (128, )
+    'in': tf.Variable(tf.constant(0.1, shape=[num_hidden_units, ])),
+    # (10, )
+    'out': tf.Variable(tf.constant(0.1, shape=[num_classes, ]))
+}
+
+
+def RNN(x, weights, biases):
+
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Use docstring

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                  "{:.3f}".format(acc))
+
+    print("Training Finished!")
+
+    # Calculate accuracy for 128 mnist test images
+    test_len = 128
+    test_data = mnist.test.images[:test_len].reshape((-1, num_step, num_input))
+    test_label = mnist.test.labels[:test_len]
+    print("Testing Accuracy:", \
+        sess.run(accuracy, feed_dict={X: test_data, Y: test_label}))
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: please use the `if __main__` etc.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    # Get lstm cell output
+    outputs, states = tf.nn.dynamic_rnn(lstm_cell, x_in, initial_state=init_state, time_major=False, dtype=tf.float32)
+
+    # Linear activation, using rnn inner loop last output
+    return tf.matmul(states[1], weights['out']) + biases['out']
+
+logits = RNN(X, weights, biases)
+prediction = tf.nn.softmax(logits)
+
+# Define loss and optimizer
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: There is too much code / definitions at the global level. please factor this

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+# Initialize the variables (i.e. assign their default value)
+init = tf.global_variables_initializer()
+
+# Start training
+with tf.Session() as sess:
+
+    # Run the initializer
+    sess.run(init)
+
+    for step in range(1, training_steps+1):
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: import `xrange` from `six` and use that.

Comment (drpngx) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Plesae use pylint

**Review by drpngx — CHANGES_REQUESTED** [review]
First pass. Mostly lint issues. You should try to run `pylint` to clear out simple style issues.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 7 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Added.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 9 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: it seems range is more recommended than xrange.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 6 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, added.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 5 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, updated.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 4 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, updated.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 3 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, updated.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 1 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, added.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 2 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Thanks, added.

**Comment by imsheridan** [comment]
@drpngx Thanks for your review, updated a new iteration. Could you please kindly have a look?

**Comment by imsheridan** [comment]
@drpngx could you please kindly have a look?

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    x_hidden = tf.reshape(x_hidden, [-1, NUM_STEP, NUM_HIDDEN_UNITS])
+
+    # Define a lstm cell with tensorflow.
+    lstm_cell = rnn.BasicLSTMCell(NUM_HIDDEN_UNITS, forget_bias=1.0, state_is_tuple=True)
+
+    # init zero state, lstm cell is consist of twp parts: (c_state, h_state).
+    init_state = lstm_cell.zero_state(BATCH_SIZE, dtype=tf.float32)
+
+    # Get lstm cell output.
+    outputs, states = tf.nn.dynamic_rnn(lstm_cell, x_hidden, \
```
Comment (drpngx) [inline tensorflow/examples/tutorials/mnist/mnist_rnn.py]: no need for '\'

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by drpngx** [comment]
I think the code is great for a tutorial. Perf and accuracy might not matter much.
/CC @MarkDaoust just checking if we're good with adding that.

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 10 above)
Comment (imsheridan) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: I assume you mean the quota in the comments, right? If so, fixed it~

**Comment by imsheridan** [comment]
@MarkDaoust could you please kindly have a check whether we're good to add this as drpngx suggested?

File: tensorflow/examples/tutorials/mnist/mnist_rnn.py (hunk 10 above)
Comment (drpngx) [reply tensorflow/examples/tutorials/mnist/mnist_rnn.py]: Sorry it didn't print correctly. I meant the backslash `\`.

**Comment by rmlarsen** [comment]
@MarkDaoust could you please take look? Thanks.

**Comment by MarkDaoust** [comment]
Hi, sorry about the delay here.

In general we're hoping to reduce the number of examples in the main repo (the [tensorflow/tensorflow/examples/](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples) directory.)

Ideally we'd put new examples over in the models repo (under [tensorflow/models/samples](https://github.com/tensorflow/models/tree/master/samples/)).

The sub directories there are:

    cookbook/ - For various api examples.
    core/ - For examples used in featured docs on the site.
    languages/ - For examples in other languages.
    outreach/ - For examples used in blogs and stuff.

We haven't totally nailed down the directory structure of "cookbook" there yet. Let's copy [api_guides](https://www.tensorflow.org/api_guides/python/contrib.bayesflow.monte_carlo), and use a flat layout with dotted name paths to each module.

So, can you make a directory for this example in [tensorflow/models](https://github.com/tensorflow/models) under `tensorflow/models/samples/nn.rnn_cell/mnist_rnn/` and we'll go from there?

**Comment by imsheridan** [comment]
Move this PR to [models](https://github.com/tensorflow/models), here is the PR: [3262](https://github.com/tensorflow/models/pull/3262). @MarkDaoust could you please help take a look?
